src/fobnail-attester.c

```diff
@@ -181,314 +87,585 @@ UsefulBuf encode_meta(struct meta_data *meta)
-        printf("Cannot create CoAP server endpoint based on UDP.\n");
-        goto error;
+    /* Create CoAP context and a client session */
+    if (!(ctx = coap_new_context(NULL))) {
+        coap_log(LOG_EMERG, "Cannot create libcoap context\n");
+        goto finish;
+    }
+    /* Support large responses */
+    coap_context_set_block_mode(ctx,
+                  COAP_BLOCK_USE_LIBCOAP | COAP_BLOCK_SINGLE_BODY);
+
+    if (!(session = coap_new_client_session(ctx, NULL, &dst,
+                                                  COAP_PROTO_UDP))) {
+        coap_log(LOG_EMERG, "Cannot create client session\n");
+        goto finish;
     }
 
-    /* register CoAP resource and resource handler */
-    printf("Registering CoAP resources.\n");
-    att_coap_add_resource(coap_context, COAP_REQUEST_FETCH, "ek", coap_ek_handler);
-    att_coap_add_resource(coap_context, COAP_REQUEST_FETCH, "aik", coap_aik_handler);
-    att_coap_add_resource(coap_context, COAP_REQUEST_FETCH, "metadata", coap_metadata_handler);
-    att_coap_add_resource(coap_context, COAP_REQUEST_FETCH, "rim", coap_rim_handler);
-    att_coap_add_resource(coap_context, COAP_REQUEST_POST, "challenge", coap_challenge_handler);
-    att_coap_add_resource(coap_context, COAP_REQUEST_POST, "quote", coap_quote_handler);
+    coap_session_set_app_data(session, &pc);
 
-    /* enter main loop */
-    printf("Entering main loop.\n");
-    while (!quit) {
-        /* process CoAP I/O */
-        if (coap_io_process(coap_context, COAP_IO_WAIT) == -1) {
-            printf("Error during CoAP I/O processing.\n");
-            goto error;
-        }
+    result = ek_request(ctx, session);
+    if (result != COAP_RESPONSE_CODE_CREATED) {
+        coap_log(LOG_EMERG, "Unexpected response for EK request (%s)\n",
+                 (result > 0 && coap_response_phrase((uint8_t)result)) ?
+                 coap_response_phrase((uint8_t)result) : "not a CoAP error");
+        goto finish;
     }
 
-    result = EXIT_SUCCESS;
-    goto finish;
+    result = aik_request(ctx, session);
+    if (result != COAP_RESPONSE_CODE_CREATED) {
+        coap_log(LOG_EMERG, "Unexpected response for AIK request (%s)\n",
+                 (result > 0 && coap_response_phrase((uint8_t)result)) ?
+                 coap_response_phrase((uint8_t)result) : "not a CoAP error");
+        goto finish;
+    }
 
-error:
-    result = EXIT_FAILURE;
+    result = provision_request(ctx, session);
+    if (result != COAP_RESPONSE_CODE_CREATED) {
+        coap_log(LOG_EMERG, "Unexpected response for provision request (%s)\n",
+                 (result > 0 && coap_response_phrase((uint8_t)result)) ?
+                 coap_response_phrase((uint8_t)result) : "not a CoAP error");
+        goto finish;
+    }
 
-finish:
-    /* free CoAP memory */
-    coap_free_endpoint(coap_endpoint);
-    coap_endpoint = NULL;
-    coap_free_context(coap_context);
-    coap_context = NULL;
+    result = meta_request(ctx, session);
```

Comment:
This may give a false impression that attester sent the request even if it didn't (such as when attester failed to obtain metadata)

Reply:
In such case `encode_sign_meta()` returns `NULLUsefulBuf`, which is caught by `meta_request()` and result is set to -1. I assume this comment applies to message printed by `coap_log()` few lines further?

Reply:
Yes, it prints: Unexpected response for metadata request